src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,181 @@
 import java.util.List;
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+    private Node<T> head;
+    private Node<T> tail;
+    private int size;
+
     @Override
     public void add(T value) {
+        Node<T> newNode = new Node<>(head, value, tail);
+        if (tail == null) {
+            head = tail = newNode;
+        } else {
+            tail.next = newNode;
+            newNode.prev = tail;
+            tail = newNode;
+        }
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        validateAddIndex(index);
+        Node<T> newNode = new Node<>(null, value, null);
+        if (head == null) {
+            head = tail = newNode;
+        } else if (index == 0) {
+            newNode.next = head;
+            head.prev = newNode;
+            head = newNode;
+        } else if (index == size) {
+            newNode.prev = tail;
+            tail.next = newNode;
+            tail = newNode;
+        } else {
+            Node<T> temporaryNode = getNode(index);
+            newNode.next = temporaryNode;
+            newNode.prev = temporaryNode.prev;
+            temporaryNode.prev.next = newNode;
+            temporaryNode.prev = newNode;
+        }
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T element : list) {
+            add(element);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        validateIndex(index);
+        Node<T> currentNode = head;
+        int counter = 0;
+        while (currentNode != null && counter != index) {
+            currentNode = currentNode.next;
+            counter++;
+        }
+        return currentNode.element;
     }
 
     @Override
     public T set(T value, int index) {
```

Comment:
You are iterating over all nodes from head to find the node at the given index. Consider iterating only until the middle of the list based on the index value to improve performance.